src/color-extractor.ts

```diff
@@ -0,0 +1,364 @@
+  colorsToIgnore: Array<string>, threshold: number): Array<any> {
+  console.log(colorsToIgnore);
+  let dataMap, keyArray, total;
+  [dataMap, keyArray, total] = getDataMap(myData);
+
+  let centerCandidates = getCenterCandidates(dataMap, keyArray, total);
+
+  let centers = new Array<string>();
+  let centerCounts, centerToColor;
+  while (centers.length === 0) {
+    [centers, centerCounts, centerToColor] = kMeans(numClusters, dataMap, centerCandidates);
+    console.log(centers);
+    [centers, centerCounts] = filterColors(centers, centerCounts, centerToColor, 
+      colorsToIgnore, threshold);
+    console.log(centers);
+  }
+
+  return [centers, centerCounts];
+}
+
+/**
+ * Gets center candidates for performing k means on the data in dataMap.
+ * Eliminates candidates if the count for a certain color is too low.
+ * 
+ * @param dataMap   Map from color to count of color in image.
+ * @param keyArray  Array containing keys of dataMap in arbitrary order.
+ * @param total     Total number of pixels from image used to make dataMap.
+ */
+function getCenterCandidates(dataMap: Map<string, number>, 
+  keyArray: Array<string>, total: number) {
+  let centerCandidates = [];
+  const threshold = total / keyArray.length / 2 ;
+  for (let i = 0; i < keyArray.length; i++) {
+    const count = dataMap.get(keyArray[i]);
+    if (count && count > threshold){
+      centerCandidates.push(keyArray[i]);
+    }
+  }
+  return centerCandidates;
+}
+
+/**
+ * Filters out cluster centers that are too similar to the specified color.
+ * Also converts center color to nearest color found in the image.
+ * Returns array of valid centers and center counts for those centers.
+ * 
+ * @param centers         Array of colors that are centers.
+ * @param centerCounts    Map from center to count of number of pixels.
+ * @param centerToColor   Map from center to colors in center's cluster.
+ */
+function filterColors(centers: Array<string>,
+  centerCounts: Array<number>,
+  centerToColor: Map<string, Array<string>>,
+  colorsToIgnore: Array<string>, threshold: number): Array<any> {
+  console.log(colorsToIgnore);
+  for (let i = centers.length - 1; i >= 0; i--) {
+    let deleteThis = false;
+    for (let j = 0; j < colorsToIgnore.length && !deleteThis; j++) {
+      if (colorDistance(colorsToIgnore[j], centers[i]) <= threshold) {
+        deleteThis = true;
+      }
+    }
+    if (!deleteThis) {
+      centers[i] = getImageColors(centerToColor, centers[i]);
+    } else {
+      centers.splice(i, 1);
+      centerCounts.splice(i, 1);
+    }
+  }
+  return [centers, centerCounts]
+  }
+
+/**
+ * Converts an ImageData object into a map from colors to counts in the 
+ * image. Also returns an array of all colors and the total number of pixels
+ * in the image. 
+ */
+function getDataMap(myData: ImageData): Array<any> {
+  let dataMap = new Map<string, number>();
+  let keyArray: Array<string> = [];
+  let total = 0;
+  for (let i = 0; i < myData.data.length; i += 4) {
```

Comment:
Re i += 4. It's not easy for readers without context to know 4 refers to the total number of color channels. Consider defining 4 as a constant with an informative name, e.g. numColorChannels.